x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/MetricsRecordingAuthenticator.java

```diff
@@ -0,0 +1,46 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.security.authc;
+
+import org.elasticsearch.telemetry.metric.LongCounter;
+
+import java.util.Map;
+import java.util.Objects;
+
+/**
+ * This class aims to simplify implementations of {@link Authenticator}s which support authentication metrics recording.
+ */
+abstract class MetricsRecordingAuthenticator implements Authenticator {
```

Comment:
I decided to change the approach and  introduce `MetricsRecordingAuthenticator` in order to avoid code duplication when adding new metrics to other `Authenticator` implementations.

Reply:
I've seen a similar implementation before that instead of an abstract class uses a wrapper to add metrics. In this case it would have the benefit of decoupling the metrics reporting from the auth code, make it unit testable and it could potentially be reused for other auth stuff with minimal effort. We could even rely on the `name` implementation to build the metrics name (not sure if this is acceptable everywhere?) and force the caller to provide a way to extract the important context fields for extra metrics label. The downside is like you pointed out in one of your comments, the added overhead of another object. 

I'm thinking something like this (not tested): 
```java
class AuthMetricsWrapper {
    public static final String METRIC_SUCCESS_COUNT = "es.security.authc.%s.success.count";
    public static final String METRIC_FAILURES_COUNT = "es.security.authc.%s.failures.count";

    static Authenticator wrap(
        Authenticator authenticator,
        Function<Authenticator.Context, Map<String, Object>> contextExtractor,
        MeterRegistry meterRegistry
    ) {
        String authProviderName = authenticator.name().toLowerCase().replace(" ", "_");
        LongCounter successCounter = meterRegistry.registerLongCounter(
            String.format(METRIC_SUCCESS_COUNT, authProviderName),
            String.format("Number of successful %s authentications.", authenticator.name()),
            "count"
        );
        LongCounter failureCounter = meterRegistry.registerLongCounter(
            String.format(METRIC_FAILURES_COUNT, authProviderName),
            String.format("Number of failed %s authentications.", authenticator.name()),
            "count"
        );
        return new Authenticator() {
            @Override
            public String name() {
                return authenticator.name();
            }

            @Override
            public AuthenticationToken extractCredentials(Context context) {
                return authenticator.extractCredentials(context);
            }

            @Override
            public void authenticate(Context context, ActionListener<AuthenticationResult<Authentication>> listener) {
                authenticator.authenticate(context, ActionListener.wrap(result -> {
                    if (result.isAuthenticated()) {
                        successCounter.incrementBy(1L, contextExtractor.apply(context));
                        listener.onResponse(result);
                    } else {
                        // ...
                    }
                }, exception -> {
                    // Add failure reason
                    failureCounter.incrementBy(1L, contextExtractor.apply(context));
                    listener.onFailure(exception);
                }));
            }
        };
    }
}
```

Reply:
Agreed. I had a very [similar idea](https://github.com/elastic/elasticsearch/pull/103178#discussion_r1420241054) in mind, but I discarded it mostly because it adds an overhead to the authentication's hot path, and I don't have a full understanding of how big of an impact it has to performance. Hence why I decided to take the safest approach of avoiding it at the cost of tighter code coupling. 
But now, I'm willing to reconsider it or at least avoid inheritance and go with the composition approach even if we don't wrap the `ActionListener`.

Reply:
> We could even rely on the name implementation to build the metrics name (not sure if this is acceptable everywhere?)

In general, this is not a bad idea, but I'd prefer to be explicit when defining a metric name and use the same names like we have defined for authentication settings. So my proposal is to have the following metric names:

- [xpack.security.authc.api_key](https://www.elastic.co/guide/en/elasticsearch/reference/current/security-settings.html#api-key-service-settings) -> es.security.authc.api_key.*
- [xpack.security.authc.token](https://www.elastic.co/guide/en/elasticsearch/reference/current/security-settings.html#token-service-settings) -> es.security.authc.token.*
- [xpack.security.authc.realms](https://www.elastic.co/guide/en/elasticsearch/reference/current/security-settings.html#realm-settings) -> es.security.authc.realms.*
- etc

Reply:
> I think my opinion will vary based on where we expect the final state. As-is, I don't think the inheritance is warranted, the super class for the counter (IMO) does not buy us much. If we plan to also include timing (how long this thing takes) then I like a wrapper since it naturally includes the start/stop times but it depends abit on exactly what we plan to measure and can hopefully easily refactor as we add functionality. As-is I think we could make the recordSuccessfulAuthentication and recordFailedAuthentication static methods if you think it helps with readability, but for these simple counts I would advocate to just count inline with the rest of the code. However, I don't have the full context of where this heading but any strategy (vs. just doing it inline) should also preferably also work with individual realms in addition to the Authenticator interface.

Sorry, I missed your comment when I was typing mine above. 
This is a good point. We do want to measure individual times. Initially I wanted to focus only on authentication results, but since I changed the approach to use a wrapper, I decided to measure authentication times as well. 
I pushed https://github.com/elastic/elasticsearch/pull/103178/commits/efe91aca8d7533cc84334ffc84f99841b6a32fb3, which removes inheritance and introduces a generic `AuthenticationMetrics` class that can be used to simply wrap the authentication action listeners in order to record metrics. With this generic wrapper approach it's also easy to measure authentication results per realm as well.

Reply:
@jfreden @jakelandis Thank you both for suggestions and the prompt to think a bit further. I'd appreciate if you could review the most recent changes.